src/partial_file.rs

```diff
@@ -0,0 +1,132 @@
+
+pub enum PartialFileRange {
+    AllFrom(u64),
+    FromTo(u64,u64),
+    Last(u64),
+}
+
+pub struct PartialFile {
+    file: File,
+    range: PartialFileRange,
+}
+
+struct PartialContentBody {
+    pub file: File,
+    pub offset: u64,
+    pub len: u64,
+}
+
+impl PartialFile {
+    pub fn new<Range>(file: File, range: Range) -> PartialFile
+        where Range: Into<PartialFileRange> {
+        let range = range.into();
+        PartialFile {
+            file: file,
+            range: range,
+        }
+    }
+
+    /// Panics if the file doesn't exist
+    pub fn from_path<P: AsRef<Path>, Range>(path: P, range: Range) -> PartialFile
+    where Range: Into<PartialFileRange> {
+        let file = File::open(path.as_ref())
+            .expect(&format!("No such file: {}", path.as_ref().display()));
+        Self::new(file, range)
+    }
+}
+
+impl From<ByteRangeSpec> for PartialFileRange {
+    fn from(b: ByteRangeSpec) -> PartialFileRange {
+        match b {
+            ByteRangeSpec::AllFrom(from) => PartialFileRange::AllFrom(from),
+            ByteRangeSpec::FromTo(from, to) => PartialFileRange::FromTo(from, to),
+            ByteRangeSpec::Last(last) => PartialFileRange::Last(last),
+        }
+    }
+}
+
+impl From<Vec<ByteRangeSpec>> for PartialFileRange {
+    fn from(v: Vec<ByteRangeSpec>) -> PartialFileRange {
+        match v.into_iter().next() {
+            // in the case no value is in "Range", return
+            // the whole file instead of panicking
+            // Note that an empty vec should never happen,
+            // but we can never be too sure
+            None => PartialFileRange::AllFrom(0),
+            Some(byte_range) => PartialFileRange::from(byte_range),
+        }
+    }
+}
+
+impl Modifier<Response> for PartialFile {
+    #[inline]
+    fn modify(self, res: &mut Response) {
+        use self::PartialFileRange::*;
+        let metadata : Option<_> = self.file.metadata().ok();
+        let file_length : Option<u64> = metadata.map(|m| m.len());
+        let range : Option<(u64, u64)> = match (self.range, file_length) {
+            (FromTo(from, to), Some(file_length)) => {
+                if from <= to && to < file_length {
+                    Some((from, to))
+                } else {
+                    None
+                }
+            },
+            (AllFrom(from), Some(file_length)) => {
+                if from < file_length {
+                    Some((from, file_length - 1))
+                } else {
+                    None
+                }
+            },
+            (Last(last), Some(file_length)) => {
+                if last < file_length {
+                    Some((file_length - last, file_length - 1))
+                } else {
+                    Some((0, file_length - 1))
+                }
+            },
+            (_, None) => None,
+            
+        };
+        if let Some(range) = range {
+            let content_range = ContentRange(ContentRangeSpec::Bytes {
```

Comment:
Setting content range can be done at the very beginning to deduplicate the code here. If the file length is `None` we can then set `RangeNotSatisfiable` and return early.